### PR TITLE
feat(arith): add TransitiveComparisonAnalyzer for inequality propagation

### PR DESCRIPTION
**Pull request by Hzfengsy**
## Summary
- Add `TransitiveComparisonAnalyzer` that builds a graph of known comparisons and derives transitive conclusions (e.g., given `x < y` and `y < z`, derive `x < z`)
- Move `CompareResult` enum to `analyzer.h` with TVM-compatible bitwise encoding, enabling intersection of comparison outcomes via `&`
- Integrate into `Analyzer::CanProve` as a fallback when bounds analysis is inconclusive
- Extract `constraint_extract.h` as a shared AND-chain decomposition utility

## Testing
- 27 new tests in `test_transitive_comparison.py` covering standalone usage, Bind, EnterConstraint, transitive chains (LE/LT/GE/EQ), offsets, Analyzer integration, and edge cases
- All 406 existing + new arith tests pass
- Full test suite (3032 tests) passes
- clang-tidy clean

## Related Issues
Part of #668 (arith simplifier tracking issue) — PR 8: TransitiveComparisonAnalyzer

### CODE REVIEW

File: src/ir/arith/constraint_extract.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  result.push_back(expr);
+  auto and_node = As<And>(expr);
+  if (and_node) {
+    auto left_parts = ExtractConstraints(and_node->left_);
+    auto right_parts = ExtractConstraints(and_node->right_);
+    result.insert(result.end(), left_parts.begin(), left_parts.end());
+    result.insert(result.end(), right_parts.begin(), right_parts.end());
+  }
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/constraint_extract.h:47]: Fixed in `a0ea6e6c` — converted to iterative worklist implementation.

File: tests/ut/ir/arith/test_transitive_comparison.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+    def test_bind_range_single_value(self):
+        t = TransitiveComparisonAnalyzer()
+        v = make_var("v")
+        t.bind(v, 5, 6)  # exactly 5
+        assert t.try_compare(v, ci(5)) == CompareResult.kEQ
+
+    def test_bind_range_bounds(self):
+        # x is already bound to [-100, 100) on the shared tca
+        assert tca.try_compare(x, ci(0)) in (CompareResult.kGE, CompareResult.kUnknown)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/arith/test_transitive_comparison.py]: Fixed in `a0ea6e6c` — tightened to `== CompareResult.kUnknown`.

File: tests/ut/ir/arith/test_transitive_comparison.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    def test_compare_with_offset(self):
+        """x <= y + 3 implies x < y + 4."""
+        y_plus_3 = ir.Add(y, ci(3), INT, S)
+        recovery = tca.enter_constraint(ir.Le(x, y_plus_3, BOOL, S))
+
+        # x compared to y+3 should give kLE
+        assert tca.try_compare(x, y_plus_3) == CompareResult.kLE
+        # x compared to y+4: we know x <= y+3, i.e. x < y+4
+        y_plus_4 = ir.Add(y, ci(4), INT, S)
+        assert tca.try_compare(x, y_plus_4) in (CompareResult.kLT, CompareResult.kLE)
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/arith/test_transitive_comparison.py]: Fixed in `a0ea6e6c` — tightened to `== CompareResult.kLT`.

File: src/ir/arith/transitive_comparison_analyzer.cpp [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+  }
+}
+
+std::function<void()> TransitiveComparisonAnalyzer::Impl::EnterConstraint(const ExprPtr& constraint) {
+  size_t old_size = scoped_knowns_.size();
+  AddKnown(constraint, &scoped_knowns_);
+  size_t new_size = scoped_knowns_.size();
+
+  if (old_size == new_size) {
+    return {};  // No constraints extracted — no cleanup needed.
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/transitive_comparison_analyzer.cpp]: Fixed in `a0ea6e6c` — now returns `[]() {}` no-op instead of `{}`.

File: python/bindings/modules/arith.cpp [hunk 5]
```diff
@@ -159,6 +158,43 @@ void BindArith(nb::module_& m) {
       .def("enter_constraint", &ir::arith::ModularSetAnalyzer::EnterConstraint, nb::arg("constraint"),
            "Enter a constraint scope. Returns a recovery function, or None if constraint is not useful.");
 
+  // CompareResult enum
+  nb::enum_<ir::arith::CompareResult>(arith, "CompareResult",
+                                      "Result of comparing two expressions.\n\n"
+                                      "Uses bitwise encoding: bit0=EQ, bit1=LT, bit2=GT.")
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/bindings/modules/arith.cpp]: Fixed in `a0ea6e6c` — added `nb::is_arithmetic()`.

File: tests/ut/ir/arith/test_transitive_comparison.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+x = make_var("x")
+y = make_var("y")
+z = make_var("z")
+a = make_var("a")
+b = make_var("b")
+c = make_var("c")
+d = make_var("d")
+for _v in [x, y, z, a, b, c, d]:
+    tca.bind(_v, -100, 100)
+
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/arith/test_transitive_comparison.py:43]: The shared `tca` pattern was an explicit design choice. All constraint-using tests properly enter/exit scopes.

File: python/pypto/pypto_core/arith.pyi [hunk 7]
```diff
@@ -10,11 +10,27 @@
...
+    """
+
+    kInconsistent = 0
+    kEQ = 1
+    kLT = 2
+    kLE = 3
+    kGT = 4
+    kGE = 5
+    kNE = 6
+    kUnknown = 7
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/pypto_core/arith.pyi]: Fixed in `a0ea6e6c` — changed to `IntFlag`.

File: python/pypto/pypto_core/arith.pyi [hunk 8]
```diff
@@ -183,6 +199,32 @@ class ModularSetAnalyzer:
...
+    @overload
+    def bind(self, var: Var, expr: Expr, allow_override: bool = False) -> None: ...
+    @overload
+    def bind(self, var: Var, min_val: int, max_val_exclusive: int, allow_override: bool = False) -> None: ...
+    def bind(self, var: Var, *args, **kwargs) -> None:
+        """Bind a variable to an expression or half-open range [min_val, max_val_exclusive)."""
+        ...
+
+    def enter_constraint(self, constraint: Expr) -> Callable[[], None]:
+        """Enter a constraint scope. Returns a recovery function that restores original state."""
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/pypto_core/arith.pyi:229]: Fixed in `a0ea6e6c` — C++ now always returns a callable (no-op when no facts added), so the stub correctly types it as `Callable[[], None]`.

File: src/ir/arith/transitive_comparison_analyzer.cpp [hunk 9]
```diff
@@ -0,0 +1,669 @@
...
+
+  /// Merge a set of comparisons sharing the same LHS/RHS into a single result.
+  [[nodiscard]] static CompareResult MergeComparisons(const std::vector<Comparison>& lhs_to_rhs,
+                                                      int64_t offset);
+
+  /// Get a canonical "zero" expression for constant offset handling.
+  ExprPtr GetZeroExpr() const;
+
+  /// Map from expression pointer to Key.
+  std::unordered_map<const Expr*, Key> expr_to_key_;
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/transitive_comparison_analyzer.cpp:175]: Fixed in `a0ea6e6c` — added `key_exprs_` vector to keep ExprPtr alive.

File: src/ir/arith/transitive_comparison_analyzer.cpp [hunk 10]
```diff
@@ -0,0 +1,669 @@
...
+    for (const auto& known : knowns_) {
+      if (auto cmp = known.WithLHS(middle_key)) {
+        attempt_transitive(cmp.value());
+      }
+    }
+    for (const auto& known : scoped_knowns_) {
+      if (auto cmp = known.WithLHS(middle_key)) {
+        attempt_transitive(cmp.value());
+      }
+    }
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/transitive_comparison_analyzer.cpp:498]: Fixed in `a0ea6e6c` — DFS now re-enqueues nodes when facts strengthen. Adjacency index deferred to follow-up if profiling shows need.

File: python/pypto/pypto_core/arith.pyi [hunk 11]
```diff
@@ -183,6 +199,32 @@ class ModularSetAnalyzer:
...
+        ...
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply python/pypto/pypto_core/arith.pyi:230]: Fixed in `a0ea6e6c` — C++ now returns no-op callable, so stub stays `Callable[[], None]` (always callable).

File: tests/ut/ir/arith/test_transitive_comparison.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+    def test_bind_range_single_value(self):
+        t = TransitiveComparisonAnalyzer()
+        v = make_var("v")
+        t.bind(v, 5, 6)  # exactly 5
+        assert t.try_compare(v, ci(5)) == CompareResult.kEQ
+
+    def test_bind_range_bounds(self):
+        # x is already bound to [-100, 100) on the shared tca
+        assert tca.try_compare(x, ci(0)) in (CompareResult.kGE, CompareResult.kUnknown)
+
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/arith/test_transitive_comparison.py:88]: Fixed in `a0ea6e6c` — tightened to `== CompareResult.kUnknown`.

File: tests/ut/ir/arith/test_transitive_comparison.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+        recovery = tca.enter_constraint(ir.Le(x, y_plus_3, BOOL, S))
+
+        # x compared to y+3 should give kLE
+        assert tca.try_compare(x, y_plus_3) == CompareResult.kLE
+        # x compared to y+4: we know x <= y+3, i.e. x < y+4
+        y_plus_4 = ir.Add(y, ci(4), INT, S)
+        assert tca.try_compare(x, y_plus_4) in (CompareResult.kLT, CompareResult.kLE)
+        # x compared to y+2 — cannot prove
+        y_plus_2 = ir.Add(y, ci(2), INT, S)
+        assert tca.try_compare(x, y_plus_2) == CompareResult.kUnknown
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply tests/ut/ir/arith/test_transitive_comparison.py:196]: Fixed in `a0ea6e6c` — tightened to `== CompareResult.kLT`.

File: include/pypto/ir/arith/analyzer.h [hunk 14]
```diff
@@ -218,6 +244,46 @@ class CanonicalSimplifier {
...
+  /// Enter a constraint scope. Returns a recovery function that restores original state.
+  std::function<void()> EnterConstraint(const ExprPtr& constraint);
+
+ private:
+  friend class Analyzer;
+  explicit TransitiveComparisonAnalyzer(Analyzer* parent);
+
+  class Impl;
+  std::unique_ptr<Impl> impl_;
+};
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply include/pypto/ir/arith/analyzer.h:288]: Fixed in `a0ea6e6c` — added `Unbind()` method to `TransitiveComparisonAnalyzer` that removes all knowns involving the variable's key. Wired into `Analyzer::Unbind`.

File: src/ir/arith/transitive_comparison_analyzer.cpp [hunk 15]
```diff
@@ -0,0 +1,669 @@
...
+    for (auto& prev_known : knowns) {
+      if (cmp.Implies(prev_known)) {
+        prev_known = cmp;
+        return;
+      }
+    }
+
+    // Independent knowledge — track separately.
+    knowns.push_back(cmp);
+  };
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/transitive_comparison_analyzer.cpp:436]: Fixed in `a0ea6e6c` — `declare_known` now re-inserts `cmp.rhs_` into `to_visit` whenever facts are strengthened or appended.

File: src/ir/arith/transitive_comparison_analyzer.cpp [hunk 16]
```diff
@@ -0,0 +1,669 @@
...
+  }
+
+  // Same expression with zero offset is always equal.
+  if (lhs_key.value() == rhs_key.value() && offset == 0) {
+    return CompareResult::kEQ;
+  }
+
+  auto lhs_to_rhs = propagate_inequalities ? CollectIndirectComparisons(lhs_key.value(), rhs_key.value())
+                                           : CollectDirectComparisons(lhs_key.value(), rhs_key.value());
+  return MergeComparisons(lhs_to_rhs, offset);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/transitive_comparison_analyzer.cpp:604]: Fixed in `a0ea6e6c` — added tautological check after `ExtractOffsets` that handles same-base comparisons (e.g., `x+1` vs `x+2`) before key lookup.

File: src/ir/arith/analyzer.cpp [hunk 17]
```diff
@@ -37,23 +37,26 @@ namespace arith {
...
   rewrite_simplify.Update(var, simplified);
+  transitive_cmp.Bind(var, simplified, allow_override);
 }
 
-void Analyzer::Bind(const VarPtr& var, int64_t min_val, int64_t max_val_exclusive, bool /*allow_override*/) {
+void Analyzer::Bind(const VarPtr& var, int64_t min_val, int64_t max_val_exclusive, bool allow_override) {
   CHECK(max_val_exclusive > min_val) << "Bind requires max_val_exclusive > min_val, got [" << min_val << ", "
                                      << max_val_exclusive << ")";
   const_int_bound.Bind(var, min_val, max_val_exclusive);
+  transitive_cmp.Bind(var, min_val, max_val_exclusive, allow_override);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply src/ir/arith/analyzer.cpp:59]: Fixed in `a0ea6e6c` — `Analyzer::Unbind` now calls `transitive_cmp.Unbind(var)`.
